File: 301-400/371-380/378-kthSmallestElementInSortedMatrix/kthSmallestElementInSortedMatrix.cpp
#include <vector>
#include <queue>
#include <functional>

using namespace std;

typedef pair<int, int> QueueItem;

template <class T>
struct Comp {
    bool operator() (const T& lhs, const T& rhs) const{
        return lhs.first > rhs.first;
    }
};

class Solution {
public:
    int kthSmallest(vector<vector<int>>& matrix, int k) {
        priority_queue<QueueItem, vector<QueueItem>, Comp<QueueItem>> queue;
        size_t size = matrix.size();
        vector<int> indices = vector<int>(size, 0);

        for (int i = 0; i < size; ++i)
            queue.push({matrix[i][0], i});

        for (int i = 1; i < k; ++i) {
            int index = queue.top().second;
            queue.pop();

            if (++indices[index] < size)
                queue.push({matrix[index][indices[index]], index});
        }

        return queue.top().first;
    }
};

File: 301-400/371-380/378-kthSmallestElementInSortedMatrix/kthSmallestElementInSortedMatrix.py
import heapq

class Solution(object):
    def kthSmallest(self, matrix, k):
        """
        :type matrix: List[List[int]]
        :type k: int
        :rtype: int
        """

        n_rows = len(matrix)
        n_cols = len(matrix[0])
        array = []
        heapq.heappush(array, (matrix[0][0], 0, 0))
        count = 0

        while array and count < k:
            value, x, y = heapq.heappop(array)
            if y < n_cols - 1:
                heapq.heappush(array, (matrix[x][y+1], x, y+1))
            if y == 0 and x < n_rows - 1:
                heapq.heappush(array, (matrix[x+1][y], x+1, y))

            count += 1

        return value